File: OsmAnd-java/src/main/java/com/google/openlocationcode/OpenLocationCode.java
package com.google.openlocationcode;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;

/**
 * Representation of open location code. https://github.com/google/open-location-code The
 * OpenLocationCode class is a wrapper around String value {@code code}, which guarantees that the
 * value is a valid Open Location Code.
 *
 * @author Jiri Semecky
 */
public final class OpenLocationCode {

  private static final BigDecimal BD_0 = new BigDecimal(0);
  private static final BigDecimal BD_5 = new BigDecimal(5);
  private static final BigDecimal BD_4 = new BigDecimal(4);
  private static final BigDecimal BD_20 = new BigDecimal(20);
  private static final BigDecimal BD_90 = new BigDecimal(90);
  private static final BigDecimal BD_180 = new BigDecimal(180);
  private static final double LATITUDE_PRECISION_8_DIGITS = computeLatitudePrecision(8) / 4;
  private static final double LATITUDE_PRECISION_6_DIGITS = computeLatitudePrecision(6) / 4;
  private static final double LATITUDE_PRECISION_4_DIGITS = computeLatitudePrecision(4) / 4;

  private static final char[] ALPHABET = "23456789CFGHJMPQRVWX".toCharArray();
  private static final Map<Character, Integer> CHARACTER_TO_INDEX = new HashMap<>();

  static {
    int index = 0;
    for (char character : ALPHABET) {
      char lowerCaseCharacter = Character.toLowerCase(character);
      CHARACTER_TO_INDEX.put(character, index);
      CHARACTER_TO_INDEX.put(lowerCaseCharacter, index);
      index++;
    }
  }

  private static final char SEPARATOR = '+';
  private static final char SEPARATOR_POSITION = 8;
  private static final char SUFFIX_PADDING = '0';

  /** Class providing information about area covered by Open Location Code. */
  public class CodeArea {

    private final BigDecimal southLatitude;
    private final BigDecimal westLongitude;
    private final BigDecimal northLatitude;
    private final BigDecimal eastLongitude;

    public CodeArea(
        BigDecimal southLatitude,
        BigDecimal westLongitude,
        BigDecimal northLatitude,
        BigDecimal eastLongitude) {
      this.southLatitude = southLatitude;
      this.westLongitude = westLongitude;
      this.northLatitude = northLatitude;
      this.eastLongitude = eastLongitude;
    }

    public double getSouthLatitude() {
      return southLatitude.doubleValue();
    }

    public double getWestLongitude() {
      return westLongitude.doubleValue();
    }

    public double getLatitudeHeight() {
      return northLatitude.subtract(southLatitude).doubleValue();
    }

    public double getLongitudeWidth() {
      return eastLongitude.subtract(westLongitude).doubleValue();
    }

    public double getCenterLatitude() {
      return southLatitude.add(northLatitude).doubleValue() / 2;
    }

    public double getCenterLongitude() {
      return westLongitude.add(eastLongitude).doubleValue() / 2;
    }

    public double getNorthLatitude() {
      return northLatitude.doubleValue();
    }

    public double getEastLongitude() {
      return eastLongitude.doubleValue();
    }
  }


  /** The state of the OpenLocationCode. */
  private final String code;

  /** Creates Open Location Code for the provided code. */
  public OpenLocationCode(String code) {
    if (!isValidCode(code)) {
      throw new IllegalArgumentException(
          "The provided code '" + code + "' is not a valid Open Location Code.");
    }
    this.code = code.toUpperCase();
  }

  /** Creates Open Location Code from the provided latitude, longitude and desired code length. */
  public OpenLocationCode(double latitude, double longitude, int codeLength)
      throws IllegalArgumentException {
    if (codeLength < 4 || (codeLength < 10 & codeLength % 2 == 1)) {
      throw new IllegalArgumentException("Illegal code length " + codeLength);
    }

    latitude = clipLatitude(latitude);
    longitude = normalizeLongitude(longitude);

    // Latitude 90 needs to be adjusted to be just less, so the returned code can also be decoded.
    if (latitude == 90) {
      latitude = latitude - 0.9 * computeLatitudePrecision(codeLength);
    }

    StringBuilder codeBuilder = new StringBuilder();

    // Ensure the latitude and longitude are within [0, 180] and [0, 360) respectively.
    /* Note: double type can't be used because of the rounding arithmetic due to floating point
     * implementation. Eg. "8.95 - 8" can give result 0.9499999999999 instead of 0.95 which
     * incorrectly classify the points on the border of a cell.
     */
    BigDecimal remainingLongitude = new BigDecimal(longitude + 180);
    BigDecimal remainingLatitude = new BigDecimal(latitude + 90);

    // Create up to 10 significant digits from pairs alternating latitude and longitude.
    int generatedDigits = 0;

    while (generatedDigits < codeLength) {
      // Always the integer part of the remaining latitude/longitude will be used for the following
      // digit.
      if (generatedDigits == 0) {
        // First step World division: Map <0..400) to <0..20) for both latitude and longitude.
        remainingLatitude = remainingLatitude.divide(BD_20);
        remainingLongitude = remainingLongitude.divide(BD_20);
      } else if (generatedDigits < 10) {
        remainingLatitude = remainingLatitude.multiply(BD_20);
        remainingLongitude = remainingLongitude.multiply(BD_20);
      } else {
        remainingLatitude = remainingLatitude.multiply(BD_5);
        remainingLongitude = remainingLongitude.multiply(BD_4);
      }
      int latitudeDigit = remainingLatitude.intValue();
      int longitudeDigit = remainingLongitude.intValue();
      if (generatedDigits < 10) {
        codeBuilder.append(ALPHABET[latitudeDigit]);
        codeBuilder.append(ALPHABET[longitudeDigit]);
        generatedDigits += 2;
      } else {
        codeBuilder.append(ALPHABET[4 * latitudeDigit + longitudeDigit]);
        generatedDigits += 1;
      }
      remainingLatitude = remainingLatitude.subtract(new BigDecimal(latitudeDigit));
      remainingLongitude = remainingLongitude.subtract(new BigDecimal(longitudeDigit));
      if (generatedDigits == SEPARATOR_POSITION) {
        codeBuilder.append(SEPARATOR);
      }
    }
    if (generatedDigits < SEPARATOR_POSITION) {
      for (; generatedDigits < SEPARATOR_POSITION; generatedDigits++) {
        codeBuilder.append(SUFFIX_PADDING);
      }
      codeBuilder.append(SEPARATOR);
    }
    this.code = codeBuilder.toString();
  }

  /** Creates Open Location Code with code length 10 from the provided latitude, longitude. */
  public OpenLocationCode(double latitude, double longitude) {
    this(latitude, longitude, 10);
  }

  public String getCode() {
    return code;
  }

  /**
   * Encodes latitude/longitude into 10 digit Open Location Code. This method is equivalent to
   * creating the OpenLocationCode object and getting the code from it.
   */
  public static String encode(double latitude, double longitude) {
    return new OpenLocationCode(latitude, longitude).getCode();
  }

  /**
   * Encodes latitude/longitude into Open Location Code of the provided length. This method is
   * equivalent to creating the OpenLocationCode object and getting the code from it.
   */
  public static String encode(double latitude, double longitude, int codeLength) {
    return new OpenLocationCode(latitude, longitude, codeLength).getCode();
  }

  /**
   * Decodes {@link OpenLocationCode} object into {@link CodeArea} object encapsulating
   * latitude/longitude bounding box.
   */
  public CodeArea decode() {
    if (!isFullCode(code)) {
      throw new IllegalStateException(
          "Method decode() could only be called on valid full codes, code was " + code + ".");
    }
    String decoded = code.replaceAll("[0+]", "");
    // Decode the lat/lng pair component.
    BigDecimal southLatitude = BD_0;
    BigDecimal westLongitude = BD_0;

    int digit = 0;
    double latitudeResolution = 400, longitudeResolution = 400;

    // Decode pair.
    while (digit < decoded.length()) {
      if (digit < 10) {
        latitudeResolution /= 20;
        longitudeResolution /= 20;
        southLatitude =
            southLatitude.add(
                new BigDecimal(latitudeResolution * CHARACTER_TO_INDEX.get(decoded.charAt(digit))));
        westLongitude =
            westLongitude.add(
                new BigDecimal(
                    longitudeResolution * CHARACTER_TO_INDEX.get(decoded.charAt(digit + 1))));
        digit += 2;
      } else {
        latitudeResolution /= 5;
        longitudeResolution /= 4;
        southLatitude =
            southLatitude.add(
                new BigDecimal(
                    latitudeResolution * (CHARACTER_TO_INDEX.get(decoded.charAt(digit)) / 4)));
        westLongitude =
            westLongitude.add(
                new BigDecimal(
                    longitudeResolution * (CHARACTER_TO_INDEX.get(decoded.charAt(digit)) % 4)));
        digit += 1;
      }
    }
    return new CodeArea(
        southLatitude.subtract(BD_90),
        westLongitude.subtract(BD_180),
        southLatitude.subtract(BD_90).add(new BigDecimal(latitudeResolution)),
        westLongitude.subtract(BD_180).add(new BigDecimal(longitudeResolution)));
  }

  /**
   * Decodes code representing Open Location Code into {@link CodeArea} object encapsulating
   * latitude/longitude bounding box.
   *
   * @param code Open Location Code to be decoded.
   * @throws IllegalArgumentException if the provided code is not a valid Open Location Code.
   */
  public static CodeArea decode(String code) throws IllegalArgumentException {
    return new OpenLocationCode(code).decode();
  }

  /** Returns whether this {@link OpenLocationCode} is a full Open Location Code. */
  public boolean isFull() {
    return code.indexOf(SEPARATOR) == SEPARATOR_POSITION;
  }

  /** Returns whether the provided Open Location Code is a full Open Location Code. */
  public static boolean isFull(String code) throws IllegalArgumentException {
    return new OpenLocationCode(code).isFull();
  }

  /** Returns whether this {@link OpenLocationCode} is a short Open Location Code. */
  public boolean isShort() {
    return code.indexOf(SEPARATOR) >= 0 && code.indexOf(SEPARATOR) < SEPARATOR_POSITION;
  }

  /** Returns whether the provided Open Location Code is a short Open Location Code. */
  public static boolean isShort(String code) throws IllegalArgumentException {
    return new OpenLocationCode(code).isShort();
  }

  /**
   * Returns whether this {@link OpenLocationCode} is a padded Open Location Code, meaning that it
   * contains less than 8 valid digits.
   */
  private boolean isPadded() {
    return code.indexOf(SUFFIX_PADDING) >= 0;
  }

  /**
   * Returns whether the provided Open Location Code is a padded Open Location Code, meaning that it
   * contains less than 8 valid digits.
   */
  public static boolean isPadded(String code) throws IllegalArgumentException {
    return new OpenLocationCode(code).isPadded();
  }

  /**
   * Returns short {@link OpenLocationCode} from the full Open Location Code created by removing
   * four or six digits, depending on the provided reference point.  It removes as many digits as
   * possible.
   */
  public OpenLocationCode shorten(double referenceLatitude, double referenceLongitude) {
    if (!isFull()) {
      throw new IllegalStateException("shorten() method could only be called on a full code.");
    }
    if (isPadded()) {
      throw new IllegalStateException("shorten() method can not be called on a padded code.");
    }

    CodeArea codeArea = decode();
    double latitudeDiff = Math.abs(referenceLatitude - codeArea.getCenterLatitude());
    double longitudeDiff = Math.abs(referenceLongitude - codeArea.getCenterLongitude());

    if (latitudeDiff < LATITUDE_PRECISION_8_DIGITS && longitudeDiff < LATITUDE_PRECISION_8_DIGITS) {
      return new OpenLocationCode(code.substring(8));
    }
    if (latitudeDiff < LATITUDE_PRECISION_6_DIGITS && longitudeDiff < LATITUDE_PRECISION_6_DIGITS) {
      return new OpenLocationCode(code.substring(6));
    }
    if (latitudeDiff < LATITUDE_PRECISION_4_DIGITS && longitudeDiff < LATITUDE_PRECISION_4_DIGITS) {
      return new OpenLocationCode(code.substring(4));
    }
    throw new IllegalArgumentException(
        "Reference location is too far from the Open Location Code center.");
  }

  /**
   * Returns an {@link OpenLocationCode} object representing a full Open Location Code from this
   * (short) Open Location Code, given the reference location.
   */
  public OpenLocationCode recover(double referenceLatitude, double referenceLongitude) {
    if (isFull()) {
      // Note: each code is either full xor short, no other option.
      return this;
    }
    referenceLatitude = clipLatitude(referenceLatitude);
    referenceLongitude = normalizeLongitude(referenceLongitude);

    int digitsToRecover = 8 - code.indexOf(SEPARATOR);
    // The resolution (height and width) of the padded area in degrees.
    double paddedAreaSize = Math.pow(20, 2 - (digitsToRecover / 2));

    // Use the reference location to pad the supplied short code and decode it.
    String recoveredPrefix =
        new OpenLocationCode(referenceLatitude, referenceLongitude)
            .getCode()
            .substring(0, digitsToRecover);
    OpenLocationCode recovered = new OpenLocationCode(recoveredPrefix + code);
    CodeArea recoveredCodeArea = recovered.decode();
    double recoveredLatitude = recoveredCodeArea.getCenterLatitude();
    double recoveredLongitude = recoveredCodeArea.getCenterLongitude();

    // Move the recovered latitude by one resolution up or down if it is too far from the reference.
    double latitudeDiff = recoveredLatitude - referenceLatitude;
    if (latitudeDiff > paddedAreaSize / 2) {
      recoveredLatitude -= paddedAreaSize;
    } else if (latitudeDiff < -paddedAreaSize / 2) {
      recoveredLatitude += paddedAreaSize;
    }

    // Move the recovered longitude by one resolution up or down if it is too far from the
    // reference.
    double longitudeDiff = recoveredCodeArea.getCenterLongitude() - referenceLongitude;
    if (longitudeDiff > paddedAreaSize / 2) {
      recoveredLongitude -= paddedAreaSize;
    } else if (longitudeDiff < -paddedAreaSize / 2) {
      recoveredLongitude += paddedAreaSize;
    }

    return new OpenLocationCode(
        recoveredLatitude, recoveredLongitude, recovered.getCode().length() - 1);
  }

  /**
   * Returns whether the bounding box specified by the Open Location Code contains provided point.
   */
  public boolean contains(double latitude, double longitude) {
    CodeArea codeArea = decode();
    return codeArea.getSouthLatitude() <= latitude
        && latitude < codeArea.getNorthLatitude()
        && codeArea.getWestLongitude() <= longitude
        && longitude < codeArea.getEastLongitude();
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    OpenLocationCode that = (OpenLocationCode) o;
    return hashCode() == that.hashCode();
  }

  @Override
  public int hashCode() {
    return code != null ? code.hashCode() : 0;
  }

  @Override
  public String toString() {
    return getCode();
  }

  // Exposed static helper methods.

  /** Returns whether the provided string is a valid Open Location code. */
  public static boolean isValidCode(String code) {
    if (code == null || code.length() < 2) {
      return false;
    }

    // There must be exactly one separator.
    int separatorPosition = code.indexOf(SEPARATOR);
    if (separatorPosition == -1) {
      return false;
    }
    if (separatorPosition != code.lastIndexOf(SEPARATOR)) {
      return false;
    }

    if (separatorPosition % 2 != 0) {
      return false;
    }

    // Check first two characters: only some values from the alphabet are permitted.
    if (separatorPosition == 8) {
      // First latitude character can only have first 9 values.
      Integer index0 = CHARACTER_TO_INDEX.get(code.charAt(0));
      if (index0 == null || index0 > 8) {
        return false;
      }

      // First longitude character can only have first 18 values.
      Integer index1 = CHARACTER_TO_INDEX.get(code.charAt(1));
      if (index1 == null || index1 > 17) {
        return false;
      }
    }

    // Check the characters before the separator.
    boolean paddingStarted = false;
    for (int i = 0; i < separatorPosition; i++) {
      if (paddingStarted) {
        // Once padding starts, there must not be anything but padding.
        if (code.charAt(i) != SUFFIX_PADDING) {
          return false;
        }
        continue;
      }
      if (CHARACTER_TO_INDEX.keySet().contains(code.charAt(i))) {
        continue;
      }
      if (SUFFIX_PADDING == code.charAt(i)) {
        paddingStarted = true;
        // Padding can start on even character: 2, 4 or 6.
        if (i != 2 && i != 4 && i != 6) {
          return false;
        }
        continue;
      }
      return false;  // Illegal character.
    }

    // Check the characters after the separator.
    if (code.length() > separatorPosition + 1) {
      if (paddingStarted) {
        return false;
      }
      // Only one character after separator is forbidden.
      if (code.length() == separatorPosition + 2) {
        return false;
      }
      for (int i = separatorPosition + 1; i < code.length(); i++) {
        if (!CHARACTER_TO_INDEX.keySet().contains(code.charAt(i))) {
          return false;
        }
      }
    }

    return true;
  }

  /** Returns if the code is a valid full Open Location Code. */
  public static boolean isFullCode(String code) {
    try {
      return new OpenLocationCode(code).isFull();
    } catch (IllegalArgumentException e) {
      return false;
    }
  }

  /** Returns if the code is a valid short Open Location Code. */
  public static boolean isShortCode(String code) {
    try {
      return new OpenLocationCode(code).isShort();
    } catch (IllegalArgumentException e) {
      return false;
    }
  }

  // Private static methods.

  private static double clipLatitude(double latitude) {
    return Math.min(Math.max(latitude, -90), 90);
  }

  private static double normalizeLongitude(double longitude) {
    if (longitude < -180) {
      longitude = (longitude % 360) + 360;
    }
    if (longitude >= 180) {
      longitude = (longitude % 360) - 360;
    }
    return longitude;
  }

  /**
   * Compute the latitude precision value for a given code length. Lengths <= 10 have the same
   * precision for latitude and longitude, but lengths > 10 have different precisions due to the
   * grid method having fewer columns than rows. Copied from the JS implementation.
   */
  private static double computeLatitudePrecision(int codeLength) {
    if (codeLength <= 10) {
      return Math.pow(20, Math.floor(codeLength / -2 + 2));
    }
    return Math.pow(20, -3) / Math.pow(5, codeLength - 10);
  }
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/util/OpenLocationCode.kt
// Copyright 2014 Google Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Modified by OsmAnd: a Kotlin copy of com.google.openlocationcode.OpenLocationCode 1.0.4, the
// library OsmAnd-java uses, with only the parts LocationParser calls.

package net.osmand.shared.util

import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.roundToLong

/**
 * An Open Location Code: a place as a short string of letters and digits, "8FVC9G8F+6X". A full
 * code is a place on its own; a short one, "9G8F+6X", is recovered from a place near it.
 *
 * Left out of the java library: `encode`, `shorten`, `contains`, the static forms of `decode`,
 * `isFull`, `isShort` and `isPadded`, and `isShortCode`.
 */
internal class OpenLocationCode {

	/** Coordinates of a decoded Open Location Code: the south-west corner, the size, and the center. */
	class CodeArea(
		val southLatitude: Double,
		val westLongitude: Double,
		val northLatitude: Double,
		val eastLongitude: Double,
		val length: Int
	) {
		val latitudeHeight: Double get() = northLatitude - southLatitude

		val longitudeWidth: Double get() = eastLongitude - westLongitude

		val centerLatitude: Double get() = (southLatitude + northLatitude) / 2

		val centerLongitude: Double get() = (westLongitude + eastLongitude) / 2
	}

	/** The code, upper case. */
	val code: String

	/** Throws [IllegalArgumentException] when [code] is not a valid code, full or short. */
	constructor(code: String) {
		if (!isValidCode(code.uppercase())) {
			throw IllegalArgumentException("The provided code '$code' is not a valid Open Location Code.")
		}
		this.code = code.uppercase()
	}

	/** Encodes a place into a code of [codeLength] digits. */
	constructor(latitude: Double, longitude: Double, codeLength: Int = CODE_PRECISION_NORMAL) {
		val length = min(codeLength, MAX_DIGIT_COUNT)
		if (length < PAIR_CODE_LENGTH && length % 2 == 1 || length < 4) {
			throw IllegalArgumentException("Illegal code length $length")
		}
		var lat = clipLatitude(latitude)
		val lng = normalizeLongitude(longitude)

		// Latitude 90 needs to be adjusted to be just less, so the returned code can also be decoded.
		if (lat == LATITUDE_MAX.toDouble()) {
			lat -= 0.9 * computeLatitudePrecision(length)
		}

		val revCodeBuilder = StringBuilder()

		// Compute the code.
		// This approach converts each value to an integer after multiplying it by
		// the final precision. This allows us to use only integer operations, so
		// avoiding any accumulation of floating point representation errors.

		// Multiply values by their precision and convert to positive without any
		// floating point operations.
		var latVal = (round((lat + LATITUDE_MAX) * LAT_INTEGER_MULTIPLIER * 1e6) / 1e6).toLong()
		var lngVal = (round((lng + LONGITUDE_MAX) * LNG_INTEGER_MULTIPLIER * 1e6) / 1e6).toLong()

		// Compute the grid part of the code if necessary.
		if (length > PAIR_CODE_LENGTH) {
			for (i in 0 until GRID_CODE_LENGTH) {
				val latDigit = latVal % GRID_ROWS
				val lngDigit = lngVal % GRID_COLUMNS
				val ndx = (latDigit * GRID_COLUMNS + lngDigit).toInt()
				revCodeBuilder.append(CODE_ALPHABET[ndx])
				latVal /= GRID_ROWS
				lngVal /= GRID_COLUMNS
			}
		} else {
			latVal = (latVal / GRID_ROWS.toDouble().pow(GRID_CODE_LENGTH)).toLong()
			lngVal = (lngVal / GRID_COLUMNS.toDouble().pow(GRID_CODE_LENGTH)).toLong()
		}
		// Compute the pair section of the code.
		for (i in 0 until PAIR_CODE_LENGTH / 2) {
			revCodeBuilder.append(CODE_ALPHABET[(lngVal % ENCODING_BASE).toInt()])
			revCodeBuilder.append(CODE_ALPHABET[(latVal % ENCODING_BASE).toInt()])
			latVal /= ENCODING_BASE
			lngVal /= ENCODING_BASE
			// If we are at the separator position, add the separator.
			if (i == 0) {
				revCodeBuilder.append(SEPARATOR)
			}
		}
		// Reverse the code.
		val codeBuilder = revCodeBuilder.reverse()

		// If we need to pad the code, replace some of the digits.
		if (length < SEPARATOR_POSITION) {
			for (i in length until SEPARATOR_POSITION) {
				codeBuilder[i] = PADDING_CHARACTER
			}
		}
		this.code = codeBuilder.substring(0, max(SEPARATOR_POSITION + 1, length + 1))
	}

	/**
	 * Decodes a full code into the area it stands for. Throws [IllegalStateException] on a short
	 * code, which [recover] makes full first.
	 */
	fun decode(): CodeArea {
		if (!isFullCode(code)) {
			throw IllegalStateException(
				"Method decode() could only be called on valid full codes, code was $code."
			)
		}
		// Strip padding and separator characters out of the code.
		val clean = code.replace(SEPARATOR.toString(), "").replace(PADDING_CHARACTER.toString(), "")

		// Initialise the values. We work them out as integers and convert them to doubles at the end.
		var latVal = -LATITUDE_MAX * LAT_INTEGER_MULTIPLIER
		var lngVal = -LONGITUDE_MAX * LNG_INTEGER_MULTIPLIER
		// This will be used to compute the value of each digit.
		var latPlaceVal = LAT_MSP_VALUE
		var lngPlaceVal = LNG_MSP_VALUE
		var i = 0
		while (i < min(clean.length, PAIR_CODE_LENGTH)) {
			latPlaceVal /= ENCODING_BASE
			lngPlaceVal /= ENCODING_BASE
			latVal += CODE_ALPHABET.indexOf(clean[i]) * latPlaceVal
			lngVal += CODE_ALPHABET.indexOf(clean[i + 1]) * lngPlaceVal
			i += 2
		}
		for (j in PAIR_CODE_LENGTH until min(clean.length, MAX_DIGIT_COUNT)) {
			latPlaceVal /= GRID_ROWS
			lngPlaceVal /= GRID_COLUMNS
			val digit = CODE_ALPHABET.indexOf(clean[j])
			val row = digit / GRID_COLUMNS
			val col = digit % GRID_COLUMNS
			latVal += row * latPlaceVal
			lngVal += col * lngPlaceVal
		}
		val latitudeLo = latVal.toDouble() / LAT_INTEGER_MULTIPLIER
		val longitudeLo = lngVal.toDouble() / LNG_INTEGER_MULTIPLIER
		val latitudeHi = (latVal + latPlaceVal).toDouble() / LAT_INTEGER_MULTIPLIER
		val longitudeHi = (lngVal + lngPlaceVal).toDouble() / LNG_INTEGER_MULTIPLIER
		return CodeArea(latitudeLo, longitudeLo, latitudeHi, longitudeHi, min(clean.length, MAX_DIGIT_COUNT))
	}

	/** A full code holds the separator after its first 8 digits. */
	fun isFull(): Boolean = code.indexOf(SEPARATOR) == SEPARATOR_POSITION

	/**
	 * The full code nearest to the reference place with this code's last digits; a full code is
	 * returned as it is.
	 */
	fun recover(referenceLatitude: Double, referenceLongitude: Double): OpenLocationCode {
		if (isFull()) {
			// Note: each code is either full xor short, no other option.
			return this
		}
		val refLatitude = clipLatitude(referenceLatitude)
		val refLongitude = normalizeLongitude(referenceLongitude)

		val digitsToRecover = SEPARATOR_POSITION - code.indexOf(SEPARATOR)
		// The precision (height and width) of the missing prefix in degrees.
		val prefixPrecision = ENCODING_BASE.toDouble().pow(2 - (digitsToRecover / 2))

		// Use the reference location to generate the prefix.
		val recoveredPrefix = OpenLocationCode(refLatitude, refLongitude).code.substring(0, digitsToRecover)
		// Combine the prefix with the short code and decode it.
		val recovered = OpenLocationCode(recoveredPrefix + code)
		val recoveredCodeArea = recovered.decode()
		// Work out whether the new code area is too far from the reference location. If it is, we
		// move it. It can only be out by a single precision step.
		var recoveredLatitude = recoveredCodeArea.centerLatitude
		var recoveredLongitude = recoveredCodeArea.centerLongitude

		// Move the recovered latitude by one precision up or down if it is too far from the reference,
		// unless doing so would lead to an invalid latitude.
		val latitudeDiff = recoveredLatitude - refLatitude
		if (latitudeDiff > prefixPrecision / 2 && recoveredLatitude - prefixPrecision > -LATITUDE_MAX) {
			recoveredLatitude -= prefixPrecision
		} else if (latitudeDiff < -prefixPrecision / 2 && recoveredLatitude + prefixPrecision < LATITUDE_MAX) {
			recoveredLatitude += prefixPrecision
		}

		// Move the recovered longitude by one precision up or down if it is too far from the
		// reference.
		val longitudeDiff = recoveredCodeArea.centerLongitude - refLongitude
		if (longitudeDiff > prefixPrecision / 2) {
			recoveredLongitude -= prefixPrecision
		} else if (longitudeDiff < -prefixPrecision / 2) {
			recoveredLongitude += prefixPrecision
		}

		return OpenLocationCode(recoveredLatitude, recoveredLongitude, recovered.code.length - 1)
	}

	companion object {

		/** Provides a normal precision code, approximately 14x14 meters. */
		const val CODE_PRECISION_NORMAL = 10

		/** The character set used to encode the digit values. */
		const val CODE_ALPHABET = "23456789CFGHJMPQRVWX"

		/** A separator used to break the code into two parts to aid memorability. */
		const val SEPARATOR = '+'

		/** The character used to pad codes. */
		const val PADDING_CHARACTER = '0'

		/** The number of characters to place before the separator. */
		private const val SEPARATOR_POSITION = 8

		/** The max number of digits to process in a plus code. */
		const val MAX_DIGIT_COUNT = 15

		/** Maximum code length using just lat/lng pair encoding. */
		private const val PAIR_CODE_LENGTH = 10

		/** Number of digits in the grid coding section. */
		private const val GRID_CODE_LENGTH = MAX_DIGIT_COUNT - PAIR_CODE_LENGTH

		/** The base to use to convert numbers to/from. */
		private const val ENCODING_BASE = CODE_ALPHABET.length

		/** The maximum value for latitude in degrees. */
		private const val LATITUDE_MAX = 90L

		/** The maximum value for longitude in degrees. */
		private const val LONGITUDE_MAX = 180L

		/** Number of columns in the grid refinement method. */
		private const val GRID_COLUMNS = 4

		/** Number of rows in the grid refinement method. */
		private const val GRID_ROWS = 5

		/** Value to multiple latitude degrees to convert it to an integer with the maximum encoding precision. */
		private const val LAT_INTEGER_MULTIPLIER = 8000L * 3125

		/** Value to multiple longitude degrees to convert it to an integer with the maximum encoding precision. */
		private const val LNG_INTEGER_MULTIPLIER = 8000L * 1024

		/** Value of the most significant latitude digit after it has been converted to an integer. */
		private const val LAT_MSP_VALUE = LAT_INTEGER_MULTIPLIER * ENCODING_BASE * ENCODING_BASE

		/** Value of the most significant longitude digit after it has been converted to an integer. */
		private const val LNG_MSP_VALUE = LNG_INTEGER_MULTIPLIER * ENCODING_BASE * ENCODING_BASE

		/** Whether [code] is a valid code, full or short. Letters may be of either case. */
		fun isValidCode(code: String?): Boolean {
			if (code == null || code.length < 2) {
				return false
			}
			val upper = code.uppercase()

			// There must be exactly one separator.
			val separatorPosition = upper.indexOf(SEPARATOR)
			if (separatorPosition == -1) {
				return false
			}
			if (separatorPosition != upper.lastIndexOf(SEPARATOR)) {
				return false
			}
			// There must be an even number of at most 8 characters before the separator.
			if (separatorPosition % 2 != 0 || separatorPosition > SEPARATOR_POSITION) {
				return false
			}

			// Check first two characters: only some values from the alphabet are permitted.
			if (separatorPosition == SEPARATOR_POSITION) {
				// First latitude character can only have first 9 values.
				if (CODE_ALPHABET.indexOf(upper[0]) > 8) {
					return false
				}
				// First longitude character can only have first 18 values.
				if (CODE_ALPHABET.indexOf(upper[1]) > 17) {
					return false
				}
			}

			// Check the characters before the separator.
			var paddingStarted = false
			for (i in 0 until separatorPosition) {
				if (CODE_ALPHABET.indexOf(upper[i]) == -1 && upper[i] != PADDING_CHARACTER) {
					// Invalid character.
					return false
				}
				if (paddingStarted) {
					// Once padding starts, there must not be anything but padding.
					if (upper[i] != PADDING_CHARACTER) {
						return false
					}
				} else if (upper[i] == PADDING_CHARACTER) {
					paddingStarted = true
					// Short codes cannot have padding
					if (separatorPosition < SEPARATOR_POSITION) {
						return false
					}
					// Padding can start on even character: 2, 4 or 6.
					if (i != 2 && i != 4 && i != 6) {
						return false
					}
				}
			}

			// Check the characters after the separator.
			if (upper.length > separatorPosition + 1) {
				if (paddingStarted) {
					return false
				}
				// Only one character after separator is forbidden.
				if (upper.length == separatorPosition + 2) {
					return false
				}
				for (i in separatorPosition + 1 until upper.length) {
					if (CODE_ALPHABET.indexOf(upper[i]) == -1) {
						return false
					}
				}
			}

			return true
		}

		/** Whether [code] is a valid full code. */
		fun isFullCode(code: String): Boolean =
			try {
				OpenLocationCode(code).isFull()
			} catch (e: IllegalArgumentException) {
				false
			}

		/** java's `Math.round`: 0 for NaN, where [roundToLong] throws. */
		private fun round(x: Double): Long = if (x.isNaN()) 0L else x.roundToLong()

		private fun clipLatitude(latitude: Double): Double =
			min(max(latitude, -LATITUDE_MAX.toDouble()), LATITUDE_MAX.toDouble())

		private fun normalizeLongitude(longitude: Double): Double {
			var lng = longitude
			while (lng < -LONGITUDE_MAX) {
				lng += LONGITUDE_MAX * 2
			}
			while (lng >= LONGITUDE_MAX) {
				lng -= LONGITUDE_MAX * 2
			}
			return lng
		}

		/**
		 * Compute the latitude precision value for a given code length. Lengths <= 10 have the same
		 * precision for latitude and longitude, but lengths > 10 have different precisions due to
		 * the grid method having fewer columns than rows. Copied from the JS implementation.
		 */
		private fun computeLatitudePrecision(codeLength: Int): Double {
			if (codeLength <= CODE_PRECISION_NORMAL) {
				return ENCODING_BASE.toDouble().pow(floor((codeLength / -2 + 2).toDouble()))
			}
			return ENCODING_BASE.toDouble().pow(-3) / GRID_ROWS.toDouble().pow(codeLength - PAIR_CODE_LENGTH)
		}
	}
}
